core/src/main/scala/quasar/plugin/postgres/CsvUpsertSink.scala

```diff
@@ -0,0 +1,196 @@
+      run(xa, writeMode))
+  }
+
+  def run[F[_]: Effect: MonadResourceErr, I](
+    xa: Transactor[F],
+    writeMode: WriteMode)(
+    args: UpsertSink.Args[F, ColumnType.Scalar, I])(
+    implicit timer: Timer[F])
+      : Stream[F, OffsetKey.Actual[I]] = {
+
+    val columns: NonEmptyList[Column[ColumnType.Scalar]] =
+      NonEmptyList(args.idColumn, args.otherColumns)
+
+    val toConnectionIO = Effect.toIOK[F] andThen LiftIO.liftK[ConnectionIO]
+
+    val table: F[Table] =
+      tableFromPath(args.path) match {
+        case Some(t) =>
+          t.pure[F]
+
+        case None =>
+          MonadResourceErr[F].raiseError(ResourceError.notAResource(args.path))
+      }
+
+
+    def handleCreate(totalBytes: Ref[F, Long], records: Chunk[Byte]): ConnectionIO[Unit] =
+      for {
+        tbl <- toConnectionIO(table)
+
+        colSpecs <- toConnectionIO(specifyColumnFragments[F](columns))
+
+        indexColumn = Fragments.parentheses(Fragment.const(hygienicIdent(args.idColumn.name)))
+
+        _ <- (args.writeMode, writeMode) match {
+          case (QWriteMode.Replace, WriteMode.Create) =>
+            createTable(log)(tbl, colSpecs) >> createIndex(log)(tbl, indexColumn)
+          case (QWriteMode.Replace, WriteMode.Replace) =>
+            dropTableIfExists(log)(tbl) >> createTable(log)(tbl, colSpecs) >> createIndex(log)(tbl, indexColumn)
+          case (QWriteMode.Replace, WriteMode.Truncate) =>
+            createTableIfNotExists(log)(tbl, colSpecs) >> truncateTable(log)(tbl) >> createIndex(log)(tbl, indexColumn)
+          case (QWriteMode.Replace, WriteMode.Append) =>
+            createTableIfNotExists(log)(tbl, colSpecs) >> createIndex(log)(tbl, indexColumn)
+          case (QWriteMode.Append, _) =>
+            ().pure[ConnectionIO]
+        }
+
+        cols = columns.map(c => hygienicIdent(c.name)).intercalate(", ")
+
+        copyQuery =
+          s"COPY ${hygienicIdent(tbl)} ($cols) FROM STDIN WITH (FORMAT csv, HEADER FALSE, ENCODING 'UTF8')"
+
+        data = records.toBytes
+
+        copied = PFCM.copyIn(copyQuery).bracketCase(
+          PFCM.embed(_, PFCI.writeToCopy(data.values, data.offset, data.length))) { (pgci, exitCase) =>
+          PFCM.embed(pgci, exitCase match {
+            case ExitCase.Completed => PFCI.endCopy.void
+            case _ => PFCI.isActive.ifM(PFCI.cancelCopy, PFCI.unit)
+          })
+        }
+
+        back <- PHC.pgGetCopyAPI(copied)
+
+        _ <- toConnectionIO(recordChunks[F](totalBytes, log)(records))
+
+      } yield back
+
+    def handleDelete(recordIds: IdBatch): ConnectionIO[Unit] =
+      if (recordIds.size === 0) {
+        ().pure[ConnectionIO]
+      } else {
+
+        def preambleFragment(tbl: Table): Fragment =
+          fr"DELETE FROM" ++
+            Fragment.const(hygienicIdent(tbl)) ++
+            fr"WHERE" ++
+            Fragment.const(hygienicIdent(args.idColumn.name))
+
+        def deleteFrom(preamble: Fragment, table: Table): ConnectionIO[Int] =
```

Comment:
Maybe I overlooked it, but I don't see `table` used anywhere.

Reply:
Fixed too.